tests/integration/cloud/end_to_end/test_pandas_datasource.py

```diff
@@ -157,3 +159,95 @@ def test_checkpoint_run(checkpoint: Checkpoint):
     """Test running a Checkpoint that was created using the entities defined in this module."""
     checkpoint_result: CheckpointResult = checkpoint.run()
     assert checkpoint_result.success
+
+
+@pytest.fixture(scope="module", params=[dataframe_asset])
+def in_memory_asset(
+    datasource: PandasDatasource,
+    request,
+) -> DataFrameAsset:
+    asset_name = f"da_{uuid.uuid4().hex}"
+    return request.param(
+        datasource=datasource,
+        asset_name=asset_name,
+    )
+
+
+@pytest.mark.cloud
+def test_checkpoint_run_in_memory(
+    cloud_base_url: str,
+    cloud_organization_id: str,
+    cloud_access_token: str,
+    context: CloudDataContext,
+    in_memory_asset: DataFrameAsset,
+    pandas_test_df: pd.DataFrame,
+    datasource: PandasDatasource,
+    expectation_suite: ExpectationSuite,
+):
+    """This Checkpoint only has one in-memory validation configured.
+    This means with a deserialized Checkpoint, if we don't update with the latest DataFrame,
+    we should get an error because nothing is there to be validated.
+    """
+    batch_request = in_memory_asset.build_batch_request(dataframe=pandas_test_df)
+    checkpoint_name = (
+        f"{in_memory_asset.name} | {expectation_suite.expectation_suite_name}"
+    )
+    validations = [
+        {
+            "expectation_suite_name": expectation_suite.expectation_suite_name,
+            "batch_request": batch_request,
+        },
+    ]
+    checkpoint: Checkpoint = context.add_checkpoint(
+        name=checkpoint_name,
+        validations=validations,
+    )
+
+    # the Checkpoint from the fixture hasn't been round-tripped,
+    # so it will work as long as everything stays in memory
+    checkpoint_result: CheckpointResult = checkpoint.run()
+    assert checkpoint_result.success
+
+    # re-retrieve the Data Context, losing the in-memory DataFrame
+    context = gx.get_context(
+        mode="cloud",
+        cloud_base_url=cloud_base_url,
+        cloud_organization_id=cloud_organization_id,
+        cloud_access_token=cloud_access_token,
+    )
+    checkpoint = context.get_checkpoint(name=checkpoint.name)
+    # failure to pass runtime validations results in error
+    with pytest.raises(RuntimeError):
+        _ = checkpoint.run()
+
+    # now that the Data Context is unaware of the Dataframe,
+    # we need to re-associate the DataFrame with the DataFrameAsset
+    # one way to do this is to assign to the attribute directly
+    # and update the datasource
+    # the fixtures came from the old Data Context
+    # we have to get them again, because they exist in a new place in memory
+    datasource = context.get_datasource(datasource_name=datasource.name)  # type: ignore[assignment]
+    in_memory_asset = datasource.get_asset(asset_name=in_memory_asset.name)
+    in_memory_asset.dataframe = pandas_test_df
+    context.update_datasource(datasource=datasource)
```

Comment:
This is weird that we have to update the datasource, but it's not documented to work by assigning to the attribute. It has to do with the logic around caching dataframes during round trips to cloud.